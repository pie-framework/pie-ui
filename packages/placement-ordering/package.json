{
  "name": "@pie-ui/placement-ordering",
  "version": "4.8.3",
  "main": "lib/index.js",
  "module": "src/index.js",
  "description": "A pie placement component",
  "keywords": [
    "pie"
  ],
  "repository": "pie-framework/pie-ui",
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "@material-ui/core": "^3.9.2",
    "@pie-lib/correct-answer-toggle": "^2.3.13",
    "@pie-lib/drag": "^1.1.45",
    "@pie-lib/math-rendering": "^2.3.1",
    "@pie-lib/render-ui": "^4.10.2",
    "debug": "^4.1.1",
    "decimal.js": "^10.0.0",
    "lodash": "^4.17.10",
    "prop-types": "^15.6.1",
    "react": "^16.8.1",
    "react-dnd": "^9.3.2",
    "react-dom": "^16.8.1",
    "react-jss": "^8.4.0",
    "react-transition-group": "^2.3.1"
  },
  "devDependencies": {
    "react-dnd-test-backend": "^2.6.0"
  },
  "gitHead": "25b2169d13da6abaaaa962e42f3f6d6ab400ec8b"
}
