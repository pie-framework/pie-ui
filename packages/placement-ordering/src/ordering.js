import cloneDeep from 'lodash/cloneDeep';
import debug from 'debug';
import isEmpty from 'lodash/isEmpty';
import map from 'lodash/map';
import assign from 'lodash/assign';

const log = debug('pie-elements:placement-ordering:ordering');

export const push = (arr, fromIndex, toIndex) => {
  const movedItem = arr.find((item, index) => index === fromIndex);
  const remainingItems = arr.filter((item, index) => index !== fromIndex);

  return [
    ...remainingItems.slice(0, toIndex),
    movedItem,
    ...remainingItems.slice(toIndex)
  ];
}

export const swap = (arr, fromIndex, toIndex) => {
  log('[swap]', arr, fromIndex, toIndex);
  if (
    !arr ||
    arr.length <= 1 ||
    fromIndex === undefined ||
    toIndex === undefined
  ) {
    throw new Error(
      `swap requires a non-empty array, fromIndex, toIndex: ${arr}, ${fromIndex} ${toIndex}`
    );
  }
  const update = cloneDeep(arr);
  const tmp = arr[toIndex];
  update[toIndex] = update[fromIndex];
  update[fromIndex] = tmp;
  return update;
};

function removeResponse(state, targetTile) {
  const update = cloneDeep(state.response);

  if (update[targetTile.index] === targetTile.id) {
    update[targetTile.index] = undefined;
  } else {
    throw new Error(
      `Tried to remove from index: ${
        targetTile.index
      }, but the id doesn't match: array: ${update}, target: ${targetTile.id}`
    );
  }
  return update;
}

function updateResponse(state, from, to) {
  const { response, opts } = state;
  const update = cloneDeep(response);
  if (opts.includeTargets) {
    if (from.type === 'choice' && to.type === 'target') {
      update[to.index] = from.id;
      return update;
    } else if (from.type === 'target' && to.type === 'choice') {
      update[from.index] = undefined;
      return update;
    } else if (from.type === 'target' && to.type === 'target') {
      return swap(response, from.index, to.index);
    } else {
      log('do nothing to the response');
      return response;
    }
  } else {
    const fromIndex = state.response.findIndex(
      r => r !== undefined && r === from.id
    );
    const toIndex = state.response.findIndex(
      r => r !== undefined && r === to.id
    );
    log('fromIndex: ', fromIndex, 'toIndex:', toIndex);
    return push(state.response, fromIndex, toIndex);
  }
}

function buildTiles(choices, response, outcomes, opts) {
  if (opts.includeTargets) {
    const targets = [];
    for (let i = 0; i < response.length; i++) {
      const r = response[i];

      const choice = choices.find(
        c => r !== undefined && r !== null && c.id === r
      );
      //TODO: index needs to match too!!
      const outcome = outcomes[i];

      const out = Object.assign(
        { type: 'target', index: i, empty: choice === undefined },
        choice,
        outcome,
        { draggable: choice !== undefined }
      );

      targets.push(out);
    }

    const processedChoices = choices.map(m => {
      if (response.indexOf(m.id) !== -1 && !opts.allowSameChoiceInTargets) {
        return {
          type: 'choice',
          empty: true,
          droppable: true,
          draggable: false
        };
      } else {
        return Object.assign({}, m, {
          type: 'choice',
          droppable: false,
          draggable: true
        });
      }
    });

    return processedChoices.concat(targets);
  } else {
    return response.reduce((acc, id, index) => {
      const emptyTile = Object.assign(
        { type: 'empty', draggable: false, droppable: true },
        choices.find(m => m.id === id),
        outcomes[index]
      );
      const choiceTile = Object.assign(
        { type: 'choice', draggable: true, droppable: true },
        choices.find(m => m.id === id),
        outcomes[index]
      )

      if(index === 0) {
        acc.push(emptyTile)
      }

      acc.push(choiceTile, emptyTile);

      return acc;
    }, []);
  }
}

export function buildState(choices, response, outcomes, opts) {
  opts = assign({ includeTargets: true }, opts);

  outcomes = outcomes || [];
  response =
    !response || isEmpty(response)
      ? opts.includeTargets
        ? new Array(choices.length)
        : map(choices, c => c.id)
      : response;
  return {
    choices,
    response,
    opts,
    outcomes,
    tiles: buildTiles(choices, response, outcomes, opts)
  };
}

export function reducer(action, state) {
  switch (action.type) {
    case 'move': {
      const { from, to } = action;
      const response = updateResponse(state, from, to);
      const tiles = buildTiles(
        state.choices,
        response,
        state.outcomes,
        state.opts
      );
      return Object.assign({}, state, { response, tiles });
    }
    case 'remove': {
      const { target } = action;
      const response = removeResponse(state, target);
      const tiles = buildTiles(
        state.choices,
        response,
        state.outcomes,
        state.opts
      );
      return Object.assign({}, state, { response, tiles });
    }
  }
  return state;
}
