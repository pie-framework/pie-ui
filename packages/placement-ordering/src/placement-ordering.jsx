import { HorizontalTiler, VerticalTiler } from './tiler';
import { buildState, reducer } from './ordering';
import { Feedback, Collapsible } from '@pie-lib/render-ui';
import CorrectAnswerToggle from '@pie-lib/correct-answer-toggle';
import PropTypes from 'prop-types';
import React from 'react';
import cloneDeep from 'lodash/cloneDeep';
import compact from 'lodash/compact';
import debug from 'debug';
import uniqueId from 'lodash/uniqueId';
import { withStyles } from '@material-ui/core/styles';

const log = debug('pie-elements:placement-ordering');

const OrderingTiler = props => {
  const {
    tiler: Comp,
    ordering,
    onDropChoice,
    onRemoveChoice,
    ...compProps
  } = props;

  return (
    <Comp
      {...compProps}
      tiles={ordering.tiles}
      onDropChoice={(t, s) => onDropChoice(t, s, ordering)}
      onRemoveChoice={t => onRemoveChoice(t, ordering)}
    />
  );
};

OrderingTiler.propTypes = {
  tiler: PropTypes.func,
  ordering: PropTypes.any,
  onDropChoice: PropTypes.func,
  onRemoveChoice: PropTypes.func
};

export class PlacementOrdering extends React.Component {
  static propTypes = {
    onSessionChange: PropTypes.func.isRequired,
    model: PropTypes.object.isRequired,
    session: PropTypes.oneOfType([
      PropTypes.array.isRequired,
      PropTypes.object.isRequired
    ]),
    classes: PropTypes.object.isRequired
  };

  constructor(props) {
    super(props);
    this.state = {
      showingCorrect: false
    };

    this.instanceId = uniqueId();

    this.toggleCorrect = showingCorrect => {
      this.setState({ showingCorrect });
    };
  }

  componentDidMount() {
    this.initSessionIfNeeded(this.props);
  }

  UNSAFE_componentWillReceiveProps(nextProps) {
    const newState = {};

    if (!nextProps.model.correctResponse) {
      newState.showingCorrect = false;
    }

    const includeTargetsChanged =
      nextProps.model.config.includeTargets !==
      this.props.model.config.includeTargets;
    const choicesNumberChanged =
      nextProps.model.choices.length !== this.props.model.choices.length;

    if (includeTargetsChanged || choicesNumberChanged) {
      this.initSessionIfNeeded(nextProps);
    }

    this.setState(newState);
  }

  initSessionIfNeeded(props) {
    const { model, session, onSessionChange } = props;
    const { config: newConfig } = model;

    const compactSessionValues = (session && compact(session.value)) || [];

    if (
      !newConfig.includeTargets &&
      compactSessionValues.length !== model.choices.length
    ) {
      log('[initSessionIfNeeded] initing session...', newConfig.includeTargets);
      const update = cloneDeep(session);

      update.value = model.choices.map(m => m.id);
      onSessionChange(update);
    } else if (newConfig.includeTargets) {
      const update = cloneDeep(session);

      delete update.value;
      onSessionChange(update);
    }
  }

  onDropChoice = (target, source, ordering) => {
    const { onSessionChange, session } = this.props;
    const from = ordering.tiles.find(
      t => t.id === source.id && t.type === source.type
    );
    const to = target;
    log('[onDropChoice] ', from, to);
    const update = reducer({ type: 'move', from, to }, ordering);
    const sessionUpdate = Object.assign({}, session, {
      value: update.response
    });

    onSessionChange(sessionUpdate);
  };

  onRemoveChoice = (target, ordering) => {
    const { onSessionChange, session } = this.props;
    log('[onRemoveChoice]', target);
    const update = reducer({ type: 'remove', target }, ordering);
    const sessionUpdate = Object.assign({}, session, {
      value: update.response
    });
    onSessionChange(sessionUpdate);
  };

  createOrdering = () => {
    const { model, session } = this.props;
    const { showingCorrect } = this.state;
    const config = model.config || {
      orientation: 'vertical',
      includeTargets: true
    };
    const { includeTargets } = config;

    return showingCorrect
      ? buildState(
          model.choices,
          model.correctResponse,
          model.correctResponse.map(id => ({ id, outcome: 'correct' })),
          {
            includeTargets,
            allowSameChoiceInTargets: model.config.allowSameChoiceInTargets
          }
        )
      : buildState(model.choices, session.value, model.outcomes, {
          includeTargets,
          allowSameChoiceInTargets: model.config.allowSameChoiceInTargets
        });
  };

  render() {
    const { classes, model } = this.props;
    const showToggle =
      model.correctResponse && model.correctResponse.length > 0;
    const { showingCorrect } = this.state;
    const config = model.config || {
      orientation: 'vertical',
      includeTargets: true
    };
    const { orientation, includeTargets } = config;
    const vertical = orientation === 'vertical';
    const ordering = this.createOrdering();

    const Tiler = vertical ? VerticalTiler : HorizontalTiler;

    return (
      <div className={classes.placementOrdering}>
        {
          model.teacherInstructions && (
            <Collapsible
              labels={{ hidden: 'Show Teacher Instructions', visible: 'Hide Teacher Instructions' }}
              className={classes.collapsible}
            >
              <div dangerouslySetInnerHTML={{ __html: model.teacherInstructions }}/>
            </Collapsible>
          )
        }
        <br />
        <CorrectAnswerToggle
          show={showToggle}
          toggled={this.state.showingCorrect}
          onToggle={this.toggleCorrect}
        />
        <div
          className={classes.prompt}
          dangerouslySetInnerHTML={{ __html: model.prompt }}
        />
        <OrderingTiler
          instanceId={this.instanceId}
          choiceLabel={config.choiceLabel}
          targetLabel={config.targetLabel}
          ordering={ordering}
          tiler={Tiler}
          disabled={model.disabled}
          addGuide={model.config.showOrdering}
          tileSize={model.config && model.config.tileSize}
          includeTargets={includeTargets}
          onDropChoice={this.onDropChoice}
          onRemoveChoice={this.onRemoveChoice}
        />
        <br />
        {
          model.rationale && (
            <Collapsible
              labels={{ hidden: 'Show Rationale', visible: 'Hide Rationale' }}
              className={classes.collapsible}
            >
              <div dangerouslySetInnerHTML={{ __html: model.rationale }}/>
            </Collapsible>
          )
        }
        {!showingCorrect && (
          <Feedback correctness={model.correctness} feedback={model.feedback} />
        )}
      </div>
    );
  }
}

const styles = theme => ({
  placementOrdering: {
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center',
    boxSizing: 'border-box'
  },
  prompt: {
    padding: '5px',
    paddingBottom: '15px'
  },
  collapsible: {
    paddingTop: theme.spacing.unit * 2,
    paddingBottom: theme.spacing.unit * 2,
    alignSelf: 'flex-start'
  }
});

export default withStyles(styles)(PlacementOrdering);
