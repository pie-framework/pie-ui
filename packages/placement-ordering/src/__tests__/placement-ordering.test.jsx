import { shallow } from 'enzyme';

import React from 'react';
import { PlacementOrdering, Choice } from '../placement-ordering';
import { buildState, reducer } from '../ordering';

jest.mock('../ordering', () => ({
  buildState: jest.fn().mockReturnValue({}),
  reducer: jest.fn().mockReturnValue({})
}));

jest.mock('@pie-lib/render-ui', () => ({
  hasText: jest.fn().mockReturnValue('true'),
  color: {
    secondary: jest.fn().mockReturnValue('blue'),
    secondaryLight: jest.fn().mockReturnValue('blue'),
    text: jest.fn().mockReturnValue('black'),
    background: jest.fn().mockReturnValue('blue'),
    incorrect: jest.fn().mockReturnValue('red'),
    correct: jest.fn().mockReturnValue('green'),
  }
}));

describe('PlacementOrdering', () => {
  let wrapper, model, session;
  let choices;
  let correctResponse;
  let onSessionChange;
  const mkWrapper = (mod, session) => {
    onSessionChange = jest.fn();
    session = { value: [], ...session };
    choices = [
      {
        id: 'c1',
        label: 'C1'
      },
      {
        id: 'c2',
        label: 'C2'
      },
      {
        id: 'c3',
        label: 'C3'
      },
      {
        id: 'c4',
        label: 'C4'
      }
    ];
    correctResponse = ['c1', 'c2', 'c3', 'c4'];
    model = {
      config: {
        includeTargets: true
      },
      choices,
      ...mod
    };

    return shallow(
      <PlacementOrdering
        model={model}
        session={session}
        classes={{
          placementOrdering: 'placementOrdering'
        }}
        onSessionChange={onSessionChange}
      />,
      {}
    );
  };

  beforeEach(() => {
    wrapper = mkWrapper();
  });

  describe('render', () => {
    it('snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('shows toggle', () => {
      let w = mkWrapper({ correctResponse: ['c1', 'c2', 'c3', 'c4'] });
      expect(w).toMatchSnapshot();
    });

    it('snapshot with rationale', () => {
      let w = mkWrapper({ rationale: 'This is rationale.' });

      expect(w).toMatchSnapshot();
    });
  });

  const ordering = d => ({
    opts: {},
    ...d
  });

  describe('logic', () => {
    let response;

    describe('onDropChoice', () => {
      beforeEach(() => {
        response = ['c4', 'c2', 'c3', 'c1'];

        reducer.mockReturnValue({ response });
        wrapper
          .instance()
          .onDropChoice(
            { id: 'c4', type: 'choice' },
            { id: 'c1', type: 'choice' },
            ordering({ tiles: [{ id: 'c1', type: 'choice' }] })
          );
      });

      it('calls reducer', () =>
        expect(reducer).toHaveBeenCalledWith(
          {
            type: 'move',
            from: { id: 'c1', type: 'choice' },
            to: { id: 'c4', type: 'choice' }
          },
          expect.anything()
        ));

      it('dropping choices updates the order', () => {
        expect(onSessionChange).toHaveBeenCalledWith({ value: response });
      });
    });

    describe('onRemoveChoice', () => {
      let target;
      beforeEach(() => {
        reducer.mockReset();
        reducer.mockReturnValue({ response: ['c1', 'c2', 'c3'] });
        wrapper = mkWrapper();

        target = { id: 'c4', type: 'target', index: 3 };
        wrapper.instance().onRemoveChoice(target, {});
      });

      it('calls reducer', () => {
        expect(reducer).toHaveBeenCalledWith(
          { type: 'remove', target },
          expect.anything()
        );
      });

      it('calls onSessionChange', () => {
        expect(onSessionChange).toHaveBeenCalledWith({
          value: ['c1', 'c2', 'c3']
        });
      });
    });

    describe('createOrdering', () => {
      let opts = { includeTargets: true, allowSameChoiceInTargets: undefined };

      beforeEach(() => {
        wrapper = mkWrapper({
          correctResponse
        });
      });

      describe('when showingCorrect is false', () => {
        beforeEach(() => {
          wrapper.setState({ showingCorrect: false });
          wrapper.instance().createOrdering();
        });

        it('calls buildState with default values', () => {
          expect(buildState).toHaveBeenCalledWith(choices, [], undefined, opts);
        });

        it('calls buildState with session values', () => {
          wrapper.setProps({ session: { value: ['c4', 'c2', 'c3', 'c1'] }});

          wrapper.instance().createOrdering();

          expect(buildState).toHaveBeenCalledWith(choices, ['c4', 'c2', 'c3', 'c1'], undefined, opts);
        });
      });

      describe('when showingCorrect is true', () => {
        let outcomes;

        beforeEach(() => {
          outcomes = correctResponse.map(id => ({ id, outcome: 'correct' }));

          wrapper.setState({ showingCorrect: true });
          wrapper.instance().createOrdering();
        });

        it('calls buildState with default values', () => {
          expect(buildState).toHaveBeenCalledWith(choices, correctResponse, outcomes, opts);
        });
      });
    });

    describe('initSessionIfNeeded', () => {
      describe('with targets', () => {
        it('resets session', () => {
          wrapper.instance().initSessionIfNeeded(wrapper.instance().props);

          expect(onSessionChange).toHaveBeenCalledWith({});
        })
      });

      describe('without targets', () => {
        let value;
        let config;

        beforeEach(() => {
          config = {
            includeTargets: false
          };

          value = choices.map(m => m.id);
          wrapper = mkWrapper({ config });

          wrapper.instance().initSessionIfNeeded(wrapper.instance().props);
        });

        it('calls onSessionChange', () => {
          expect(onSessionChange).toHaveBeenCalledWith({ value });
        });

        it('does not call onSessionChange if session is set', () => {
          wrapper = mkWrapper({ config }, { value });

          wrapper.instance().initSessionIfNeeded(wrapper.instance().props);

          expect(onSessionChange).not.toBeCalled();
        });
      });
    });

    describe('UNSAFE_componentWillReceiveProps', () => {
      beforeEach(() => {
        wrapper = mkWrapper({
          config: {
            includeTargets: false
          }
        });
      });

      it('calls initSessionIfNeeded if includeTargets changes', () => {
        wrapper.setProps({
          model: {
            ...model,
            config: {
              includeTargets: true
            }
          }
        });

        expect(onSessionChange).toHaveBeenCalledWith({});
      });

      it('calls initSessionIfNeeded if choicesNumberChanged changes', () => {
        wrapper.setProps({
          model: {
            ...model,
            choices: choices.slice(0, 2)
          }
        });

        expect(onSessionChange).toHaveBeenCalledWith({ value: ['c1', 'c2']});
      });
    });
  });
});
