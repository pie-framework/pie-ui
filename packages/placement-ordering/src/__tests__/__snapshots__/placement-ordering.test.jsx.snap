// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlacementOrdering render shows toggle 1`] = `
<div
  className="placementOrdering"
>
  <br />
  <_default
    onToggle={[Function]}
    show={true}
    toggled={false}
  />
  <div
    dangerouslySetInnerHTML={
      Object {
        "__html": undefined,
      }
    }
  />
  <OrderingTiler
    includeTargets={true}
    instanceId="3"
    onDropChoice={[Function]}
    onRemoveChoice={[Function]}
    ordering={Object {}}
    tiler={[Function]}
  />
  <br />
  <Feedback />
</div>
`;

exports[`PlacementOrdering render snapshot 1`] = `
<div
  className="placementOrdering"
>
  <br />
  <_default
    onToggle={[Function]}
    toggled={false}
  />
  <div
    dangerouslySetInnerHTML={
      Object {
        "__html": undefined,
      }
    }
  />
  <OrderingTiler
    includeTargets={true}
    instanceId="1"
    onDropChoice={[Function]}
    onRemoveChoice={[Function]}
    ordering={Object {}}
    tiler={[Function]}
  />
  <br />
  <Feedback />
</div>
`;

exports[`PlacementOrdering render snapshot with rationale 1`] = `
<div
  className="placementOrdering"
>
  <br />
  <_default
    onToggle={[Function]}
    toggled={false}
  />
  <div
    dangerouslySetInnerHTML={
      Object {
        "__html": undefined,
      }
    }
  />
  <OrderingTiler
    includeTargets={true}
    instanceId="5"
    onDropChoice={[Function]}
    onRemoveChoice={[Function]}
    ordering={Object {}}
    tiler={[Function]}
  />
  <br />
  <Component
    labels={
      Object {
        "hidden": "Show Rationale",
        "visible": "Hide Rationale",
      }
    }
  >
    <div
      dangerouslySetInnerHTML={
        Object {
          "__html": "This is rationale.",
        }
      }
    />
  </Component>
  <Feedback />
</div>
`;
